KIPs/kip-247.md

```diff
@@ -0,0 +1,198 @@
+  - (otherwise) `amountRepay = amountRepay(GaslessSwapTx)`.
+  - `amountRepay(GaslessApproveTx, GaslessSwapTx) = V1 + V2 + V3` and `amountRepay(GaslessSwapTx) = V1 + V3` where
+    - `V1 = LendTx.Fee() = SwapTx.GasPrice() * 21000`
+    - `V2 = ApproveTx.Fee()` (if exists)
+    - `V3 = SwapTx.Fee()`
+
+While GaslessApproveTx must be followed by with GaslessSwapTx in order to be promoted, GaslessSwapTx can exist by itself.
+
+### Definition of LendTx
+
+Upon detection of a GaslessTx (either GaslessApproveTx or GaslessSwapTx), the block proposer shall inject a transaction, denoted by LendTx, which transfers KAIA to the GaslessTx sender. LendTx must meet all the following requirements:
+
+- L1: `LendTx.type` is 0x7802, a Ethereum dynamic fee transaction type.
+- L2: `LendTx.from` is the proposer itself.
+- L3: `LendTx.to` is `GaslessSwapTx.from`
+- L4: `LendTx.value` is `amountLend`
+  - `amountLend` is `V2 + V3` (from aforementioned `amountRepay` definition).
+
+### GaslessSwapRouter
+
+GaslessSwapRouter is a smart contract responsible for token swap and repayment. It supports only single-hop swaps through a predetermined DEX contract. It must support the following interface:
+
+```
+interface IKIP247 {
+    function swapForGas(address token, uint256 amountIn, uint256 minAmountOut, uint256 amountRepay) external;
+    function addToken(address token, address dex) external;
+    function removeToken(address token) external;
+
+    function claimCommission() external;
+    function updateCommissionRate(uint256 _commissionRate) external;
+
+    // view functions
+    function dexAddress(address token) external view returns (address dex);
+    function getAmountIn(address token, uint amountOut) external view returns (uint amountIn);
+}
+```
+
+`swapForGas()` must ensure that all the following conditions are satisfied:
+
+- R1: Sender has enough tokens. That is, `token.balanceOf(msg.sender) >= amountIn`.
+- R2: Token is whitelisted and has a corresponding DEX address other than zero.
+- R3: `amountReceived >= minAmountOut >= amountRepay` where `amountReceived` is the swap output.
+
+`swapForGas()` must distribute the swap output (`amountReceived`) as follows:
+
+- The block proposer (`block.coinbase`) receives `amountRepay`.
+- User receives `(amountReceived - amountRepay) * (1 - commissionRate)`.
+- The commission `(amountReceived - amountRepay) * commissionRate` is stored in the contract for later claim.
+
+### TxPool Policy Updates
+
+#### GaslessTx Validation (validateTx)
+
+To prevent GaslessTx to be discarded, the validation logic at `txpool.queue` insertion needs to detect GaslessTx and skip the account balance check.
+
+```
+// txpool.validateTx()
+if ...
+else: // if tx is NOT feeDelegatedTx
+    if senderBalance.Cmp(tx.Cost()) < 0:
+        if tx is GaslessTx:
+            validation ok
+        else:
+            insufficientFund error
+```
+
+#### Promoting GaslessTx (promoteExecutables)
```

Comment:
In practice, there is a more complicated case. I provides some examples:

[transfer_0, approve_1, swap_2] is queue contents.
[transfer_0, approve_1, swap_2] shoud be promoted to pending.

[approve_0, swap_1, approve_2, swap_3] is a queue list.
[approve_0, swap_1, approve_2, swap_3] shoud be promoted to pending.

[approve_0, approve_1, swap_2] is a queue list.
[] shoud be promoted to pending.

[approve_0, transfer_1, swap_2] is a queue list.
[] shoud be promoted to pending.

Is this collect? And I think there is a lack of checking of nonce overall.

Reply:
I think we should only allow either
- (if approve exists) when getNonce(user) == GaslessApproveTx.nonce and getNonce(user) + 1 == GaslessSwapTx.nonce
- (else) when getNonce(user) == GaslessSwapTx.nonce

Thus, these are NOT regarded as GaslessTx:
- approve_1 and swap_2 in [transfer_0, approve_1, swap_2]
- approve_2 and swap_3 in [approve_0, swap_1, approve_2, swap_3]
- approve_1 and swap_2 in [approve_0, approve_1, swap_2]

The nonce check added by previous commit was meant to imply this nonce check.